### PR TITLE
Prelim dstess routine (sym trid spectrum slicer)

### PR DESCRIPTION
**Pull request by Tony-ICL**
Added new compute/dstess.c, a spectrum slicer for symmetric tri-diagonal
matrices. This reproduces the basic functionality of LAPACK dstevx,
with OpenMP parallelism. CMakeLists.txt is modified to include it.
Values were added to plasma_types.h to support it (which match the
ones found in the plasma_sevp fork). plasma_z.h has an added
prototype for plasma_dstess, and subs.py was modified to allow other
prototypes to be formed; even though we only have the double precision
form at this time. This has been tested with a standalone tester,
test_dstess.c, which is not included in the repository.

### CODE REVIEW

File: compute/dstess.c [hunk 1]
```diff
@@ -0,0 +1,1333 @@
...
+// void dorgtsqr_(int *M, int *N, int *MB, int *NB, double *A, int *LDA, double *T, int *LDT, double *WORK, int *LWORK, int *INFO);
+// void dstein_(int *N, double *D, double *E, int *M, double *W, int *IBLOCK, int *ISPLIT,
+//            double *Z, int *LDZ, double *WORK, int *IWORK, int *IFAIL, int *INFO);
+
+
+/*******************************************************************************
+ *
+ * plasma_dstess: Symmetric Tridiagonal Eigen Spectrum Slicer.
+ *
+ * Computes a caller-selected range of eigenvalues and, optionally,
```
Comment (luszczek) [inline compute/dstess.c:46]: This routine, code, and description should be modeled after LAPACK's `*TEVX()`

File: compute/dstess.c [hunk 2]
```diff
@@ -0,0 +1,1333 @@
...
+   T = malloc(LDT*N*RowB*sizeof(double));
+   // If we cannot allocate: Try to be sensible and try again.
+   // The formula divides by (MB-N) which is the culprit; MB-N is three.
+   // Force MB=M so RowB = 1.
+   if (T == NULL) {
+      MB=M;
+      RowB = 1;
+      T = malloc(LDT*N*RowB*sizeof(double));
+      if (T == NULL) {
+         DSTESS_DEBUG(0) fprintf(stderr, "%s:%i Out of memory for T, LDT*N*RowB=%i.\n", __func__, __LINE__, LDT*N*RowB);
```
Comment (luszczek) [inline compute/dstess.c]: I/O commands do not belong in PLASMA's compute  routines.

File: compute/dstess.c [hunk 3]
```diff
@@ -0,0 +1,1333 @@
...
+   double W = u;
+
+   if (myArrays[thread].IBLOCK == NULL || myArrays[thread].ISPLIT==NULL || 
+       myArrays[thread].WORK==NULL || myArrays[thread].IWORK==NULL || 
+       myArrays[thread].IFAIL==NULL) {
+      DSTESS_DEBUG(1) fprintf(stderr, "%2i:%s:%i dstein failed to allocate workspaces.\n", omp_get_thread_num(), __func__, __LINE__);
+      return(PlasmaErrorOutOfMemory);
+   }
+
+   dstein_(&N, diag, offd, &M, &W, myArrays[thread].IBLOCK, myArrays[thread].ISPLIT, v,
```
Comment (luszczek) [inline compute/dstess.c]: This call must go through offical LAPACK wrapper (not an ad hoc one).

File: compute/dstess.c [hunk 4]
```diff
@@ -0,0 +1,1333 @@
...
+   if (myArrays[thread].IBLOCK == NULL || myArrays[thread].ISPLIT==NULL || 
+       myArrays[thread].WORK==NULL || myArrays[thread].IWORK==NULL || 
+       myArrays[thread].IFAIL==NULL) {
+      DSTESS_DEBUG(1) fprintf(stderr, "%2i:%s:%i dstein failed to allocate workspaces.\n", omp_get_thread_num(), __func__, __LINE__);
+      return(PlasmaErrorOutOfMemory);
+   }
+
+   dstein_(&N, diag, offd, &M, &W, myArrays[thread].IBLOCK, myArrays[thread].ISPLIT, v, 
+           &LDZ, myArrays[thread].WORK, myArrays[thread].IWORK, myArrays[thread].IFAIL, &INFO);
+   DSTESS_DEBUG(0) fprintf(stderr, "%2i:%s:%i ev=%.16e dstein returning INFO=%d.\n",
```
Comment (luszczek) [inline compute/dstess.c]: No error reporting through I/O in compute routines.

File: compute/dstess.c [hunk 5]
```diff
@@ -0,0 +1,1333 @@
...
+// \lambda_{max} = max (i=0; i<n) diag[i]+|offd[i]| + |offd[i-1]|,
+// with offd[-1], offd[n] = 0.
+// Indexes above are 0 relative.
+// Although Gerschgorin is mentioned in ?larr?.f LAPACK files, it is coded
+// inline there. This is a simple min/max boundary for the full range of
+// eigenvalues. However, there is no guarantee that \lambda_{min} and
+// \lambda_{max} are actually eigenvalues. We add a step to find the actual min
+// and max, so we can also use this to compute the matrix Norm. 
+//-----------------------------------------------------------------------------
+static void Bound_MinMax_Eigvalue(double *diag, double *offd, int n, double *Min, double *Max) {
```
Comment (luszczek) [inline compute/dstess.c:444]: This should be modeled after LAPACK's `*larr*()` routines

File: compute/dstess.c [hunk 6]
```diff
@@ -0,0 +1,1333 @@
...
+// routine DSTEIN, storing it in its proper place in pVec.  Then we add numEV
+// to the count of finished eigenpairs, and free the bracket.
+//
+// The most comparable serial LAPACK routine is DSTEVX.
+//
+// Once all thread work is complete, the code will condense these arrays to
+// just the actual number of unique eigenvalues found.
+//-----------------------------------------------------------------------------
+
+void thread_work(WorkStack* Stack) {
```
Comment (luszczek) [inline compute/dstess.c:595]: This is an external name and it's too generic. And it's large enough for it's own name and source code file.

File: compute/dstess.c [hunk 7]
```diff
@@ -0,0 +1,1333 @@
...
+} // END MM.
+
+
+//-----------------------------------------------------------------------------
+// eigenpair error: If A*v = u*v, then A*v-u*v should == 0. We compute the
+// L_infinity norm of (A*v-u*v).
+// We return DBL_MAX if the eigenvector (v) is all zeros.
+// If u==0.0, we'll return L_INF of (A*V). 
+//-----------------------------------------------------------------------------
+static double eigp_error(double *diag, double *offd, int n, double u, double *v) {
```
Comment (luszczek) [inline compute/dstess.c:518]: This routine is probably for testing, not for user compute functionality.

File: compute/dstess.c [hunk 8]
```diff
@@ -0,0 +1,1333 @@
...
+   } // end Master Loop.
+
+   DSTESS_DEBUG(0) fprintf(stderr, "%2i:%s:%i Exiting thread work.\n", omp_get_thread_num(), __func__, __LINE__);
+} // end thread_work.
+
+
+//-----------------------------------------------------------------------------
+// This is the main routine; plasma_dstess. 
+//-----------------------------------------------------------------------------
+int plasma_dstess(plasma_enum_t eigt, plasma_enum_t range, // args 1,2
```
Comment (luszczek) [inline compute/dstess.c:805]: `DSTEVX()` should be the model for the name.

File: compute/dstess.c [hunk 9]
```diff
@@ -0,0 +1,1333 @@
...
+            free(myB);
+            break;
+      } // End switch on stage.
+   } // end Master Loop.
+
+   DSTESS_DEBUG(0) fprintf(stderr, "%2i:%s:%i Exiting thread work.\n", omp_get_thread_num(), __func__, __LINE__);
+} // end thread_work.
+
+
+//-----------------------------------------------------------------------------
```
Comment (luszczek) [inline compute/dstess.c:802]: I don't see Doxygen comments though it's main serial entry point.

File: include/plasma_types.h [hunk 10]
```diff
@@ -94,6 +94,17 @@ enum {
     PlasmaMaxNorm       = 177,
     PlasmaRealMaxNorm   = 178,
 
+    PlasmaNoVec         = 301,
```
Comment (luszczek) [inline include/plasma_types.h:97]: These should be reconciled with `SEVP` fork of PLASMA

File: include/plasma_z.h [hunk 11]
```diff
@@ -218,6 +218,17 @@ int plasma_zpotrs(plasma_enum_t uplo,
                   plasma_complex64_t *pA, int lda,
                   plasma_complex64_t *pB, int ldb);
 
+int plasma_zstess(plasma_enum_t eigt, plasma_enum_t range,
```
Comment (luszczek) [inline include/plasma_z.h:221]: I think only `DSTESS` and `SSTESS` would be needed.

**Review by luszczek — COMMENTED** [review]
This is just a first set of comments. It's best to close this PR for now. We should work on the fork directly before creating a new PR.

**Comment by Tony-ICL** [comment]
Too many corrections needed, and trying a different strategy. -Tony

**Pull request closed without merging**